--- qemu-0.9.0-o/dyngen.c	2007-02-06 14:44:57.000000000 -0700
+++ qemu-0.9.0/dyngen.c	2007-02-06 14:46:11.000000000 -0700
@@ -1692,6 +1692,9 @@
 #else
         fprintf(outfile, "    extern void %s();\n", name);
 #endif
+#if defined(HOST_PPC)
+    uint8_t *blr_addr = NULL;
+#endif
 
         for(i = 0, rel = relocs;i < nb_relocs; i++, rel++) {
             host_ulong offset = get_rel_offset(rel);
@@ -2053,6 +2056,9 @@
 #else
 #error unsupport object format
 #endif
+		if (blr_addr)
+		    fprintf(outfile, "    *(uint32_t *)(gen_code_ptr + %d) = 0x48000000 | %d;\n",
+			    blr_addr - p_start, p_end - blr_addr);
             }
 #elif defined(HOST_S390)
             {
