diff -Nur qemu-0.9.0.orig/hw/ps2.c qemu-0.9.0.chng/hw/ps2.c
--- qemu-0.9.0.orig/hw/ps2.c	2007-02-06 00:01:54.000000000 +0100
+++ qemu-0.9.0.chng/hw/ps2.c	2007-03-14 12:52:00.000000000 +0100
@@ -162,7 +162,7 @@
             q->rptr = 0;
         q->count--;
         /* reading deasserts IRQ */
-        s->update_irq(s->update_arg, 0);
+//        s->update_irq(s->update_arg, 0);
         /* reassert IRQs if data left */
         s->update_irq(s->update_arg, q->count != 0);
     }
