Index: qemu-snapshot-2007-02-09_05/hw/pc.c
===================================================================
--- qemu-snapshot-2007-02-09_05.orig/hw/pc.c
+++ qemu-snapshot-2007-02-09_05/hw/pc.c
@@ -522,15 +522,13 @@ static void pc_init1(int ram_size, int v
     cpu_register_physical_memory(0xc0000, 0x10000, 
                                  vga_bios_offset | IO_MEM_ROM);
 
-    /* map the last 128KB of the BIOS in ISA space */
+    /* copy the last 128KB of the BIOS to ISA space */
     isa_bios_size = bios_size;
     if (isa_bios_size > (128 * 1024))
         isa_bios_size = 128 * 1024;
-    cpu_register_physical_memory(0xd0000, (192 * 1024) - isa_bios_size,
-                                 IO_MEM_UNASSIGNED);
-    cpu_register_physical_memory(0x100000 - isa_bios_size,
-                                 isa_bios_size,
-                                 (bios_offset + bios_size - isa_bios_size) | IO_MEM_ROM);
+    memcpy(phys_ram_base + 0x100000 - isa_bios_size,
+           phys_ram_base + bios_offset + bios_size - isa_bios_size,
+           isa_bios_size);
 
     {
         ram_addr_t option_rom_offset;
