--- qemu-0.5.2/Makefile.target~	2004-01-25 16:50:02.000000000 +0100
+++ qemu-0.5.2/Makefile.target	2004-02-05 02:42:54.000000000 +0100
@@ -53,7 +53,7 @@
 endif # !CONFIG_USER_ONLY
 
 ifdef CONFIG_STATIC
-LDFLAGS+=-static
+#LDFLAGS+=-static
 endif
 
 ifeq ($(ARCH),i386)
@@ -97,7 +97,7 @@
 OP_CFLAGS=$(CFLAGS) -fno-delayed-branch -ffixed-i0
 HELPER_CFLAGS=$(CFLAGS) -ffixed-i0 -mflat
 # -static is used to avoid g1/g3 usage by the dynamic linker
-LDFLAGS+=-Wl,-T,$(SRC_PATH)/sparc.ld -static
+#LDFLAGS+=-Wl,-T,$(SRC_PATH)/sparc.ld -static
 endif
 
 ifeq ($(ARCH),sparc64)
@@ -211,7 +211,7 @@
 VL_LDFLAGS=
 # specific flags are needed for non soft mmu emulator
 ifdef CONFIG_STATIC
-VL_LDFLAGS+=-static
+#VL_LDFLAGS+=-static
 endif
 ifndef CONFIG_SOFTMMU
 VL_LDFLAGS+=-Wl,-T,$(SRC_PATH)/i386-vl.ld 
