--- qemu-8.2.4/include/qemu/osdep.h.orig	2024-05-13 21:19:44.000000000 +0200
+++ qemu-8.2.4/include/qemu/osdep.h	2024-05-28 21:35:11.927464569 +0200
@@ -27,10 +27,6 @@
 #ifndef QEMU_OSDEP_H
 #define QEMU_OSDEP_H
 
-#if !defined _FORTIFY_SOURCE && defined __OPTIMIZE__ && __OPTIMIZE__ && defined __linux__
-# define _FORTIFY_SOURCE 2
-#endif
-
 #include "config-host.h"
 #ifdef NEED_CPU_H
 #include CONFIG_TARGET
