--- qemu-0.7.0/dyngen.c.orig	2005-04-27 22:52:05.000000000 +0200
+++ qemu-0.7.0/dyngen.c	2005-05-01 14:00:00.000000000 +0200
@@ -1194,13 +1194,11 @@
     } else if (strstart(sym_name, "__op_gen_label", &p)) {
         snprintf(name, name_size, "gen_labels[param%s]", p);
     } else {
-#ifdef HOST_SPARC
         if (sym_name[0] == '.')
-            snprintf(name, sizeof(name),
+            snprintf(name, name_size,
                      "(long)(&__dot_%s)",
                      sym_name + 1);
         else
-#endif
             snprintf(name, name_size, "(long)(&%s)", sym_name);
     }
 }
@@ -1590,14 +1588,12 @@
                     !strstart(sym_name, "__op_param", NULL) &&
                     !strstart(sym_name, "__op_jmp", NULL) &&
                     !strstart(sym_name, "__op_gen_label", NULL)) {
-#if defined(HOST_SPARC)
 		    if (sym_name[0] == '.') {
 			fprintf(outfile,
 				"extern char __dot_%s __asm__(\"%s\");\n",
 				sym_name+1, sym_name);
 			continue;
 		    }
-#endif
 #if defined(__APPLE__)
 /* set __attribute((unused)) on darwin because we wan't to avoid warning when we don't use the symbol */
                     fprintf(outfile, "extern char %s __attribute__((unused));\n", sym_name);
