--- qemu-7.1.0/configure.orig	2022-10-05 06:29:01.214549869 +0200
+++ qemu-7.1.0/configure	2022-10-05 16:30:40.352312168 +0200
@@ -2246,10 +2246,6 @@ esac
 
 write_c_skeleton
 
-if test "$fortify_source" = "yes" ; then
-  QEMU_CFLAGS="-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 $QEMU_CFLAGS"
-fi
-
 if test "$have_asan" = "yes"; then
   QEMU_CFLAGS="-fsanitize=address $QEMU_CFLAGS"
   QEMU_LDFLAGS="-fsanitize=address $QEMU_LDFLAGS"
