--- qemu-2.5.0/configure~	2015-12-21 22:39:21.000000000 +0100
+++ qemu-2.5.0/configure	2015-12-21 22:39:58.952714685 +0100
@@ -4754,7 +4754,7 @@
 echo "GNUTLS support    $gnutls"
 echo "GNUTLS hash       $gnutls_hash"
 echo "libgcrypt         $gcrypt"
-echo "nettle            $nettle ${nettle+($nettle_version)}"
+echo "nettle            $nettle ${nettle+\($nettle_version)}"
 echo "libtasn1          $tasn1"
 echo "VTE support       $vte"
 echo "curses support    $curses"
