--- qemu-1.7.0/configure.orig	2013-11-27 23:15:55.000000000 +0100
+++ qemu-1.7.0/configure	2013-12-15 08:37:16.187664605 +0100
@@ -332,10 +332,6 @@
 QEMU_CFLAGS="-Wstrict-prototypes -Wredundant-decls $QEMU_CFLAGS"
 QEMU_CFLAGS="-D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE $QEMU_CFLAGS"
 QEMU_INCLUDES="-I. -I\$(SRC_PATH) -I\$(SRC_PATH)/include"
-if test "$debug_info" = "yes"; then
-    CFLAGS="-g $CFLAGS"
-    LDFLAGS="-g $LDFLAGS"
-fi
 
 # make source path absolute
 source_path=`cd "$source_path"; pwd`
@@ -3549,8 +3545,6 @@
 if test "$gcov" = "yes" ; then
   CFLAGS="-fprofile-arcs -ftest-coverage -g $CFLAGS"
   LDFLAGS="-fprofile-arcs -ftest-coverage $LDFLAGS"
-elif test "$debug" = "no" ; then
-  CFLAGS="-O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 $CFLAGS"
 fi
 
 
