diff -urN qemu-0.6.0-old/Makefile qemu-0.6.0/Makefile
--- qemu-0.6.0-old/Makefile	2004-07-14 17:17:57.259195608 +0000
+++ qemu-0.6.0/Makefile	2004-07-14 17:27:48.597298552 +0000
@@ -47,20 +47,20 @@
         done
 
 install: all 
-	mkdir -p "$(bindir)"
+	mkdir -p "$(DESTDIR)$(bindir)"
 ifndef CONFIG_WIN32
 	install -m 755 -s $(TOOLS) "$(bindir)"
 endif
-	mkdir -p "$(datadir)"
+	mkdir -p "$(DESTDIR)$(datadir)"
 	install -m 644 pc-bios/bios.bin pc-bios/vgabios.bin \
                        pc-bios/vgabios-cirrus.bin \
                        pc-bios/ppc_rom.bin \
-                       pc-bios/linux_boot.bin "$(datadir)"
-	mkdir -p "$(docdir)"
-	install -m 644 qemu-doc.html  qemu-tech.html "$(docdir)"
+                       pc-bios/linux_boot.bin "$(DESTDIR)$(datadir)"
+	mkdir -p "$(DESTDIR)$(docdir)"
+	install -m 644 qemu-doc.html  qemu-tech.html "$(DESTDIR)$(docdir)"
 ifndef CONFIG_WIN32
-	mkdir -p "$(mandir)/man1"
-	install qemu.1 qemu-mkcow.1 "$(mandir)/man1"
+	mkdir -p "$(DESTDIR)$(mandir)/man1"
+	install qemu.1 qemu-mkcow.1 "$(DESTDIR)$(mandir)/man1"
 endif
 	for d in $(TARGET_DIRS); do \
 	$(MAKE) -C $$d $@ || exit 1 ; \
diff -urN qemu-0.6.0-old/Makefile.target qemu-0.6.0/Makefile.target
--- qemu-0.6.0-old/Makefile.target	2004-07-14 17:17:57.287191352 +0000
+++ qemu-0.6.0/Makefile.target	2004-07-14 17:31:32.545253288 +0000
@@ -340,7 +340,7 @@
 
 install: all 
 ifneq ($(PROGS),)
-	install -m 755 -s $(PROGS) "$(bindir)"
+	install -m 755 -s $(PROGS) "$(DESTDIR)$(bindir)"
 endif
 
 ifneq ($(wildcard .depend),)
