--- qemu-0.6.0/cpu-exec.c.orig	2004-07-10 20:20:09.000000000 +0200
+++ qemu-0.6.0/cpu-exec.c	2004-10-16 15:40:11.763462432 +0200
@@ -903,7 +903,7 @@
 # define TRAP_sig(context)			EXCEPREG_sig(exception, context) /* number of powerpc exception taken */
 #endif /* __APPLE__ */
 
-int cpu_signal_handler(int host_signum, siginfo *info, 
+int cpu_signal_handler(int host_signum, struct siginfo *info, 
                        void *puc)
 {
     struct ucontext *uc = puc;
