--- qemu-0.12.2.a/configure	2010-02-02 17:41:58.000000000 +0100
+++ qemu-0.12.2.b/configure	2010-02-02 17:58:54.000000000 +0100
@@ -1267,7 +1267,7 @@
 if test "$curses" != "no" ; then
   curses_found=no
   cat > $TMPC << EOF
-#include <curses.h>
+#include <ncurses/curses.h>
 #ifdef __OpenBSD__
 #define resize_term resizeterm
 #endif
--- qemu-0.12.2.a/ui/curses.c	2010-02-02 18:29:18.000000000 +0100
+++ qemu-0.12.2.a/ui/curses.c	2010-02-02 18:29:47.000000000 +0100
@@ -21,7 +21,7 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  * THE SOFTWARE.
  */
-#include <curses.h>
+#include <ncurses/curses.h>
 
 #ifndef _WIN32
 #include <signal.h>
--- qemu-0.13.0/ui/curses_keys.h~	2010-10-15 21:56:09.000000000 +0100
+++ qemu-0.13.0/ui/curses_keys.h	2010-10-20 13:38:16.578832783 +0100
@@ -22,7 +22,7 @@
  * THE SOFTWARE.
  */
 
-#include <curses.h>
+#include <ncurses/curses.h>
 #include "keymaps.h"
 
 
--- qemu-1.0/console.h.wiget	2011-12-01 23:44:44.809570527 +0100
+++ qemu-1.0/console.h	2011-12-01 23:45:02.775490445 +0100
@@ -329,7 +329,7 @@ static inline int ds_get_bytes_per_pixel
 }
 
 #ifdef CONFIG_CURSES
-#include <curses.h>
+#include <ncurses/curses.h>
 typedef chtype console_ch_t;
 #else
 typedef unsigned long console_ch_t;
