--- qemu-0.6.0/Makefile.orig	2004-07-10 20:20:09.000000000 +0200
+++ qemu-0.6.0/Makefile	2004-10-16 12:44:46.537539984 +0200
@@ -43,21 +43,21 @@
         done
 
 install: all 
-	mkdir -p "$(bindir)"
+	mkdir -p "$(DESTDIR)$(bindir)"
 ifndef CONFIG_WIN32
-	install -m 755 -s $(TOOLS) "$(bindir)"
+	install -m 755 -s $(TOOLS) "$(DESTDIR)$(bindir)"
 endif
-	mkdir -p "$(datadir)"
+	mkdir -p "$(DESTDIR)$(datadir)"
 	install -m 644 pc-bios/bios.bin pc-bios/vgabios.bin \
                        pc-bios/vgabios-cirrus.bin \
                        pc-bios/ppc_rom.bin \
                        pc-bios/proll.bin \
-                       pc-bios/linux_boot.bin "$(datadir)"
-	mkdir -p "$(docdir)"
-	install -m 644 qemu-doc.html  qemu-tech.html "$(docdir)"
+                       pc-bios/linux_boot.bin "$(DESTDIR)$(datadir)"
+	mkdir -p "$(DESTDIR)$(docdir)"
+	install -m 644 qemu-doc.html  qemu-tech.html "$(DESTDIR)$(docdir)"
 ifndef CONFIG_WIN32
-	mkdir -p "$(mandir)/man1"
-	install qemu.1 qemu-mkcow.1 "$(mandir)/man1"
+	mkdir -p "$(DESTDIR)$(mandir)/man1"
+	install qemu.1 qemu-mkcow.1 "$(DESTDIR)$(mandir)/man1"
 endif
 	for d in $(TARGET_DIRS); do \
 	$(MAKE) -C $$d $@ || exit 1 ; \
--- qemu-0.6.0/Makefile.target.orig	2004-10-16 12:43:34.780448720 +0200
+++ qemu-0.6.0/Makefile.target	2004-10-16 12:45:08.871144760 +0200
@@ -340,7 +340,7 @@
 
 install: all 
 ifneq ($(PROGS),)
-	install -m 755 -s $(PROGS) "$(bindir)"
+	install -m 755 -s $(PROGS) "$(DESTDIR)$(bindir)"
 endif
 
 ifneq ($(wildcard .depend),)
