--- qemu-2.5.0/configure~	2015-12-21 22:27:18.000000000 +0100
+++ qemu-2.5.0/configure	2015-12-21 22:29:00.990327069 +0100
@@ -4494,12 +4494,8 @@
 # After here, no more $cc or $ld runs
 
 if test "$gcov" = "yes" ; then
-  CFLAGS="-fprofile-arcs -ftest-coverage -g $CFLAGS"
+  CFLAGS="-fprofile-arcs -ftest-coverage $CFLAGS"
   LDFLAGS="-fprofile-arcs -ftest-coverage $LDFLAGS"
-elif test "$fortify_source" = "yes" ; then
-  CFLAGS="-O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 $CFLAGS"
-elif test "$debug" = "no"; then
-  CFLAGS="-O2 $CFLAGS"
 fi
 
 ##########################################
