--- qemu-7.0.0/configure.orig	2022-10-02 20:59:57.508988228 +0200
+++ qemu-7.0.0/configure	2022-10-02 21:09:37.279180676 +0200
@@ -2705,7 +2705,7 @@ QEMU_GA_MSI_MINGW_DLL_PATH="$($pkg_confi
 
 # Mac OS X ships with a broken assembler
 roms=
-if { test "$cpu" = "i386" || test "$cpu" = "x86_64"; } && \
+if { test "$cpu" = "i386" || test "$cpu" = "x86_64" || test "$cpu" = "x32"; } && \
         test "$targetos" != "darwin" && test "$targetos" != "sunos" && \
         test "$targetos" != "haiku" && test "$softmmu" = yes ; then
     # Different host OS linkers have different ideas about the name of the ELF
