--- qemu-2.2.0/configure~	2015-03-26 19:52:14.000000000 +0100
+++ qemu-2.2.0/configure	2015-03-26 19:56:34.255072716 +0100
@@ -6148,7 +6148,7 @@
 
 # Mac OS X ships with a broken assembler
 roms=
-if { test "$cpu" = "i386" || test "$cpu" = "x86_64"; } && \
+if { test "$cpu" = "i386" || test "$cpu" = "x86_64" || test "$cpu" = "x32"; } && \
         test "$targetos" != "Darwin" && test "$targetos" != "SunOS" && \
         test "$targetos" != "Haiku" && test "$softmmu" = yes ; then
     # Different host OS linkers have different ideas about the name of the ELF
