--- qemu-2.2.0/configure~	2015-03-26 19:52:14.000000000 +0100
+++ qemu-2.2.0/configure	2015-03-26 19:56:34.255072716 +0100
@@ -1904,7 +1904,7 @@
 
 if test "$seccomp" != "no" ; then
     case "$cpu" in
-    i386|x86_64)
+    i386|x86_64|x32)
         libseccomp_minver="2.1.0"
         ;;
     arm|aarch64)
@@ -2932,7 +2932,7 @@
 ##########################################
 # TPM passthrough is only on x86 Linux
 
-if test "$targetos" = Linux && test "$cpu" = i386 -o "$cpu" = x86_64; then
+if test "$targetos" = Linux && test "$cpu" = i386 -o "$cpu" = x86_64 -o "$cpu" = x32; then
   tpm_passthrough=$tpm
 else
   tpm_passthrough=no
@@ -4161,7 +4161,7 @@
 
 # Mac OS X ships with a broken assembler
 roms=
-if test \( "$cpu" = "i386" -o "$cpu" = "x86_64" \) -a \
+if test \( "$cpu" = "i386" -o "$cpu" = "x86_64" -o "$cpu" = "x32" \) -a \
         "$targetos" != "Darwin" -a "$targetos" != "SunOS" -a \
         "$softmmu" = yes ; then
   roms="optionrom"

